File: sgi-webapp/src/app/core/models/csp/estado-proyecto.ts
import { marker } from '@biesbjerg/ngx-translate-extract-marker';
import { DateTime } from 'luxon';

export interface IEstadoProyecto {
  /** Id */
  id: number;
  /** ID del proyecto */
  proyectoId: number;
  /** Estado */
  estado: Estado;
  /** Fecha estado */
  fechaEstado: DateTime;
  /** Comentario */
  comentario: string;
}

export enum Estado {
  BORRADOR = 'BORRADOR',
  CONCEDIDO = 'CONCEDIDO',
  RENUNCIADO = 'RENUNCIADO',
  RESCINDIDO = 'RESCINDIDO',
}

export const ESTADO_MAP: Map<Estado, string> = new Map([
  [Estado.BORRADOR, marker(`csp.estado-proyecto.BORRADOR`)],
  [Estado.CONCEDIDO, marker(`csp.estado-proyecto.CONCEDIDO`)],
  [Estado.RENUNCIADO, marker(`csp.estado-proyecto.RENUNCIADO`)],
  [Estado.RESCINDIDO, marker(`csp.estado-proyecto.RESCINDIDO`)]
]);

File: sgi-csp-service/src/main/java/org/crue/hercules/sgi/csp/model/EstadoProyecto.java
package org.crue.hercules.sgi.csp.model;

import java.time.Instant;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.ForeignKey;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "estado_proyecto")
@Data
@EqualsAndHashCode(callSuper = false)
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class EstadoProyecto extends BaseEntity {

  /**
   * Serial version
   */
  private static final long serialVersionUID = 1L;

  /**
   * Estados del proyecto
   */
  public enum Estado {
    /** Borrador */
    BORRADOR,
    /** Concedido */
    CONCEDIDO,
    /** Renunciado */
    RENUNCIADO,
    /** Rescindido */
    RESCINDIDO;
  }

  /** Id */
  @Id
  @Column(name = "id", nullable = false)
  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "estado_proyecto_seq")
  @SequenceGenerator(name = "estado_proyecto_seq", sequenceName = "estado_proyecto_seq", allocationSize = 1)
  private Long id;

  /** Proyecto Id */
  @Column(name = "proyecto_id", nullable = false)
  @NotNull
  private Long proyectoId;

  /** Tipo estado proyecto */
  @Column(name = "estado", length = 50, nullable = false)
  @Enumerated(EnumType.STRING)
  @NotNull
  private Estado estado;

  /** Fecha. */
  @Column(name = "fecha_estado", nullable = false)
  @NotNull
  private Instant fechaEstado;

  /** Comentario */
  @Column(name = "comentario", length = 2000, nullable = true)
  @Size(max = 2000)
  private String comentario;

  // Relation mappings for JPA metamodel generation only
  @ManyToOne
  @JoinColumn(name = "proyecto_id", insertable = false, updatable = false, foreignKey = @ForeignKey(name = "FK_ESTADOPROYECTO_PROYECTO"))
  @Getter(AccessLevel.NONE)
  @Setter(AccessLevel.NONE)
  private final Proyecto proyecto = null;
}